### PR TITLE
FQDN: Avoid a data race on FQDN policyAdd

### PR DESCRIPTION
**Pull request by eloycoto**
This commit change a little bit how policyAdd handles the case of FQDN
rules. This code checks that the FQDN rule UUID is still present on the
fqdn.RuleGen, if it is not present means that the `daemon.policyAdd`
already updated the rule, so the rule is no longer valid.

Checking the RuleGen.allRules is O(1) due is a map, hopefully this will
be replaced in the future when policyRepo has a key-value and things can
be retrieved without looping over it.

This is a new commit,and this will close the PR#7220 where issue was
discussed more on deep.

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/7459)

### CODE REVIEW

**Comment by eloycoto** [comment]
test-me-please

File: daemon/policy.go
Comment (raybejjani) [inline daemon/policy.go]: I guess it should be ok because the queue ensures things are done one-at-a-time but I was expecting this check to happen in `policyAdd` after the policy repository lock was held. 
Wherever we do it, we should include a debug print here saying that this is a no-op. We also need to make sure that we don't need to handle returning a revision number or error (I'm not sure what the new code assumes, it may always wait for a response).

File: pkg/fqdn/rulegen.go
Comment (raybejjani) [inline pkg/fqdn/rulegen.go]: I would put this explanation in the place we call this function instead. The actual race is in how all this is connected in daemon, specifically around the order of operations when different things call `PolicyAdd`. I would go into a little more detail (I'm assuming that the check is done with the lock held, though, so the text may need to change if you don't want to do that):
This check is needed to avoid a race where a DNS update to a rule enqueues an copy of that rule that is older than a k8s/API update to that same rule. Rules are assigned random UUIDs in MarkToFQDNsRule when initially inserted and DNS updates always replace existing rules. In this out-of-order scenario the k8s/API updated rule would receive a new UUID and the older UUID, belonging to the DNS updated rule, would no longer be present. This check fails in that case, and relies on the Add/Remove of a toFQDNs rule in RuleGen happening atomically.

File: pkg/fqdn/rulegen.go
Comment (raybejjani) [inline pkg/fqdn/rulegen.go]: This comment is probably a carry-over?

File: pkg/fqdn/rulegen.go
Comment (raybejjani) [inline pkg/fqdn/rulegen.go:169]: We can `return true` here unconditionally, no? Since we return false in the loop when a rule has a toFQDNs section and it is not in RuleGen, finishing the loop means all rules were not toFQDNs, or they were all toFQDNs and in RuleGen (or some combination). That would also simplify the loop a bit since you don't need to accumulate `result` or `ruleHasFQDN`

File: pkg/fqdn/rulegen.go
Comment (raybejjani) [inline pkg/fqdn/rulegen.go]: I wanted to say that we can shortcut everything and just search for the UUID label only. We should leave a comment explaining that Labels.Get is expensive and that hasToFQDN is a quicker sanity check.

**Review by raybejjani — CHANGES_REQUESTED** [review]

File: pkg/fqdn/rulegen.go
Comment (eloycoto) [reply pkg/fqdn/rulegen.go]: I just wanted to make it sure that has a FQDN because maybe one day a UUID is added and all rules stop working ;-)

File: daemon/policy.go
Comment (eloycoto) [reply daemon/policy.go]: Around return, etc.. all are ok :-)

About the place, for me is more in the handle, something that you want to check before you add the rules in policyAdd and avoid logs, etc.. but I'm open here.

**Comment by eloycoto** [comment]
test-me-please

File: daemon/policy.go
Comment (raybejjani) [inline daemon/policy.go]: I just realised an issue with this. The DNS updates include all rules that refer to a DNS name. Some may have been replaced, others may still be there. `AreFQDNRulesStillPresent` would return false if any rule has been replaced, but the others still need to go through. This might mean we need to filter out rules that are no longer valid :/

File: pkg/fqdn/rulegen_test.go
Comment (raybejjani) [inline pkg/fqdn/rulegen_test.go:279]: Would it be possible to also test the out-of-order handling in a until test for daemon? I think we can create the actual bad situation with the queue.

**Review by raybejjani — CHANGES_REQUESTED** [review]

**Review by nebril — APPROVED** [review]

File: pkg/fqdn/rulegen_test.go
Comment (ianvernon) [reply pkg/fqdn/rulegen_test.go:279]: Is there a way we can have a field in `toFQDNs` rules that has them marked for deletion? This could be done independently of the EventQueue since the actual rule content wouldn't be modified (?)

**Comment by eloycoto** [comment]
test-me-please

**Comment by eloycoto** [comment]
test-me-please

File: pkg/fqdn/rulegen.go
Comment (tgraf) [inline pkg/fqdn/rulegen.go]: MarkToFQDNRules no longer exists

File: pkg/fqdn/rulegen.go
Comment (tgraf) [inline pkg/fqdn/rulegen.go:145]: The commit does not describe why it is safe to omit the rest of the loop in the case when the UUID has not been found. It seems like a change in behavior that the commit message does not describe.

**Review by tgraf — CHANGES_REQUESTED** [review]

**Comment by eloycoto** [comment]
test-me-please

**Comment by eloycoto** [comment]
test-me-please

**Review by tgraf — APPROVED** [review]
I understand the motivation now and the code does what is described. I'll defer to @raybejjani for whether this is correct behavior.

File: daemon/policy.go
Comment (raybejjani) [inline daemon/policy.go:250]: This might be more understandable if it gave more context:
All rules being added have ToFQDNs UUIDs that have been removed and will not be re-inserted to avoid a race.

File: pkg/fqdn/rulegen.go
Comment (raybejjani) [inline pkg/fqdn/rulegen.go:120]: Do we need to worry about the difference between a nil slice and non-nil empty? I think no but this forces it one way.

File: pkg/fqdn/rulegen_test.go
Comment (raybejjani) [reply pkg/fqdn/rulegen_test.go:279]: Do you mean the `EgressRule` in `api.Rule.Egress`? We can add it but I'm not sure if we have any private fields on there right now or if we can (since swagger make these types for us). I think you mentioned it elsewhere but we could make it public but it's unclear how long a rule would be reported in this state (since, if we could detect this race condition we wouldn't enqueue the replacement rule anyway).

Separately, how did you see the data moving about? One frustrating thing is that once an api.Rule is added to the policy repository a copy is made. This was the original reason I had to resort to the UUID label since that was the only identifier that would stick. RuleGen also happens to make a copy that it passes to PolicyAdd. We could change that (but risk introducing a race on the rule objects themselves if multiple updates are enqueued).

File: daemon/policy.go
Comment (raybejjani) [inline daemon/policy.go:246]: Could you add a comment here saying something like:
There is a small race between this PrepareFQDNRules invocation and taking the policy lock. As long as policyAdd is fed by a single-threaded queue this should never be an issue.


I'm worried we will optimise later and bad things will happen to us.

**Review by raybejjani — APPROVED** [review]

File: pkg/fqdn/rulegen.go
Comment (eloycoto) [reply pkg/fqdn/rulegen.go:120]: no, we don't need to worry about!

**Comment by eloycoto** [comment]
test-me-please

File: pkg/fqdn/rulegen_test.go
Comment (ianvernon) [reply pkg/fqdn/rulegen_test.go:279]: Yeah, there are probably a lot of small details that make what I am suggesting much more difficult. This implementation is fine :)

**Review by ianvernon — APPROVED** [review]

**Comment by tgraf** [comment]
@eloycoto @raybejjani Does this need backporting to 1.4?

**Comment by eloycoto** [comment]
I do not think so, needs the event channel to run correctly.

**Pull request merged**
